### PR TITLE
[Merged by Bors] - Add OpenTelemetry/Jaeger tracing backend

### PR DESCRIPTION
**Pull request by nicholastmosher**
This is the first step towards improving our observability within Fluvio. This adds a jaeger backend to our tracing setup, so you can now run a jaeger instance and collect traces.

- Adds Jaeger backend to tracing subscriber for SC and SPU
- Adds more tracing instrumentation to the SPU

To try this out for yourself, run Jaeger in a docker container with the following command:

```
docker run -d -p6831:6831/udp -p6832:6832/udp -p16686:16686 -p14268:14268 jaegertracing/all-in-one:latest
```

Then open up `http://localhost:16686` in your browser. Try running some Fluvio commands to interact with the cluster, then view traces sorted by `fluvio-spu` or `fluvio-sc`.

I don't know what we want the ultimate scope of this PR to be. It could potentially be merged now, but do we want to explore any of the following first?

- Add more instrumentation to SC? I think that'd be a good idea
- Add instrumentation to the client, i.e. Fluvio CLI? If yes, we would want to make sure it is optional and probably disabled by default
- Try adding prometheus metrics? There are examples showing how this works, but I believe it requires creating a metrics route on an HTTP server, so it would really only make sense for the SC and SPU. This is probably out of scope for this.

### CODE REVIEW

**Comment by sehz** [comment]
linked to issue?

File: src/protocol/Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "fluvio-protocol"
 edition = "2018"
-version = "0.5.1"
+version = "0.5.2"
```
Comment (sehz) [inline src/protocol/Cargo.toml:4]: this has not been published yet, so we can still keep 0.5.1

File: src/protocol/fluvio-protocol-derive/Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "fluvio-protocol-derive"
-version = "0.2.1"
+version = "0.2.2"
```
Comment (sehz) [inline src/protocol/fluvio-protocol-derive/Cargo.toml:3]: latest published version is `0.2.0`

File: src/protocol/fluvio-protocol-derive/src/de.rs [hunk 3]
```diff
@@ -17,6 +17,10 @@ pub(crate) fn generate_decode_trait_impls(input: &DeriveItem) -> TokenStream {
             let (impl_generics, ty_generics, where_clause) = kf_struct.generics.split_for_impl();
             quote! {
                 impl #impl_generics fluvio_protocol::Decoder for #ident #ty_generics #where_clause {
+                    #[tracing::instrument(
+                        skip(self, src, version),
+                        fields(type_name = std::any::type_name::<#ident #ty_generics>()),
+                    )]
```
Comment (sehz) [inline src/protocol/fluvio-protocol-derive/src/de.rs:23]: this is too low level to add instrument?

File: src/protocol/fluvio-protocol-derive/src/ser.rs [hunk 4]
```diff
@@ -16,12 +16,14 @@ pub(crate) fn generate_encode_trait_impls(input: &DeriveItem) -> TokenStream {
             let size_field_tokens = parse_struct_props_size(&kf_struct.props, &ident);
             quote! {
                 impl #impl_generics fluvio_protocol::Encoder for #ident #ty_generics #where_clause {
+                    #[tracing::instrument(skip(self, dest, version))]
```
Comment (sehz) [inline src/protocol/fluvio-protocol-derive/src/ser.rs:19]: same.  let's not add instrument to this crate

File: src/runner/Cargo.toml [hunk 5]
```diff
@@ -18,6 +18,12 @@ path = "src/bin/main.rs"
 doc = false
 
 [dependencies]
+tracing = "0.1"
+tracing-subscriber = "0.2"
+tracing-opentelemetry = "0.10"
+opentelemetry = "0.11"
+opentelemetry-jaeger = "0.10"
```
Comment (sehz) [inline src/runner/Cargo.toml:25]: make jaeger as optional

File: src/runner/src/bin/main.rs [hunk 6]
```diff
@@ -1,10 +1,27 @@
...
+
+    let (tracer, _uninstall) = opentelemetry_jaeger::new_pipeline()
+        .with_service_name(service_name)
+        .install()
+        .unwrap();
+
+    Registry::default()
+        .with(tracing_subscriber::EnvFilter::new("DEBUG"))
+        .with(tracing_subscriber::fmt::layer())
+        .with(tracing_opentelemetry::layer().with_tracer(tracer))
```
Comment (sehz) [inline src/runner/src/bin/main.rs:22]: make this as optional configuration.

File: src/spu/src/replication/leader/leaders_state.rs [hunk 7]
```diff
@@ -43,17 +43,20 @@ impl<S> ReplicaLeadersState<S> {
 
 impl<S> ReplicaLeadersState<S> {
     /// get clone of state
+    #[instrument(skip(self, replica))]
```
Comment (sehz) [inline src/spu/src/replication/leader/leaders_state.rs:46]: remove here

File: src/spu/src/services/public/stream_fetch.rs [hunk 8]
```diff
@@ -449,10 +449,12 @@ pub mod publishers {
         }
 
         // get next stream id
+        #[instrument(skip(self))]
         fn next_stream_id(&self) -> u32 {
             self.stream_id.fetch_add(1, SeqCst)
         }
 
+        #[instrument(skip(self))]
```
Comment (sehz) [inline src/spu/src/services/public/stream_fetch.rs:457]: remove

File: src/storage/src/replica.rs [hunk 9]
```diff
@@ -37,20 +37,24 @@ impl Unpin for FileReplica {}
 impl ReplicaStorage for FileReplica {
     type Config = ConfigOption;
 
+    #[instrument(skip(replica, config))]
     async fn create(replica: &ReplicaKey, config: Self::Config) -> Result<Self, StorageError> {
         Self::create(replica.topic.clone(), replica.partition as u32, 0, config).await
     }
 
+    #[instrument(skip(self))]
```
Comment (sehz) [inline src/storage/src/replica.rs:45]: remove

**Review by sehz — CHANGES_REQUESTED** [review]
Careful on adding instrument otherwise it will lead to generation of too many low level spans. Instrument should be added
* API call
* Long duration async, typically network or long storage (for example, sending status)

File: src/protocol/Cargo.toml (hunk 1 above)
Comment (nicholastmosher) [reply src/protocol/Cargo.toml:4]: Sure, I just needed to use this to force Cargo to compile my local changes, e.g. for fluvio-protocol-derive

File: src/protocol/fluvio-protocol-derive/src/de.rs (hunk 3 above)
Comment (nicholastmosher) [reply src/protocol/fluvio-protocol-derive/src/de.rs:23]: Part of the benefit of putting this here is that, since we manually control the Encoding/Decoding, we may want visibility into the performance of the encode/decode implementations for various pieces of data in case something is accidentally taking a long time when it shouldn't. That was my rationale at least. If you think it is too low-level I can remove it

Comment (sehz) [reply src/protocol/fluvio-protocol-derive/src/de.rs:23]: encode/decode is at `ns` resolution (this will be be compiler optimized) which is too low level which was not designed for tracing.

File: src/spu/src/services/public/stream_fetch.rs (hunk 8 above)
Comment (nicholastmosher) [reply src/spu/src/services/public/stream_fetch.rs:457]: The original reason I put this in this area because of the async mutexes involved. I was thinking it would be a good way to spot any problems if we get any high lock contention or anything like that.

File: src/runner/src/bin/main.rs (hunk 6 above)
Comment (nicholastmosher) [reply src/runner/src/bin/main.rs:22]: Hmm I am trying, but it seems like it is not working when I make things optional, even if the feature is enabled.

Comment (nicholastmosher) [reply src/runner/src/bin/main.rs]: Just figured out the solution to this, PR is updated

**Comment by nicholastmosher** [comment]
I have named the feature for enabling/disabling opentelemetry+jaeger "telemetry". Let me know if you'd prefer a different name for that, e.g. just "jaeger" or something

File: src/client/src/sockets.rs [hunk 10]
```diff
@@ -214,7 +218,9 @@ impl VersionedSerialSocket {
         // send request & save response
         self.socket.send_and_receive(req_msg).await
     }
+
     /// create new request based on version
+    #[instrument(skip(self, request, version))]
```
Comment (sehz) [inline src/client/src/sockets.rs:223]: this is low level.  either put trace or remove it

File: src/protocol/fluvio-protocol-api/src/request.rs [hunk 11]
```diff
@@ -85,10 +85,12 @@ where
     }
 
     #[allow(unused)]
+    #[instrument(skip(self, response))]
```
Comment (sehz) [inline src/protocol/fluvio-protocol-api/src/request.rs:88]: put trace mode

File: src/protocol/fluvio-protocol-api/src/request.rs [hunk 12]
```diff
@@ -85,10 +85,12 @@ where
...
     pub fn new_response(&self, response: R::Response) -> ResponseMessage<R::Response> {
         Self::response_with_header(&self.header, response)
     }
 
+    #[instrument(skip(header, response))]
```
Comment (sehz) [inline src/protocol/fluvio-protocol-api/src/request.rs:93]: trace mode

File: src/protocol/fluvio-protocol-api/src/request.rs [hunk 13]
```diff
@@ -97,6 +99,7 @@ where
     }
 
     #[allow(unused)]
+    #[instrument(skip(self, src, version))]
```
Comment (sehz) [inline src/protocol/fluvio-protocol-api/src/request.rs:102]: trace mode

File: src/protocol/fluvio-protocol-api/src/request.rs [hunk 14]
```diff
@@ -132,6 +136,7 @@ impl<R> Decoder for RequestMessage<R>
 where
     R: Request,
 {
+    #[instrument(skip(self, src, version))]
     fn decode<T>(&mut self, src: &mut T, version: Version) -> Result<(), IoError>
```
Comment (sehz) [inline src/protocol/fluvio-protocol-api/src/request.rs:140]: trace mode

File: src/protocol/fluvio-protocol-api/src/request.rs [hunk 15]
```diff
@@ -150,6 +155,7 @@ where
         self.header.write_size(version) + self.request.write_size(self.header.api_version())
     }
 
+    #[instrument(skip(self, out, version))]
```
Comment (sehz) [inline src/protocol/fluvio-protocol-api/src/request.rs:158]: trace mode

File: src/runner/Cargo.toml [hunk 16]
```diff
@@ -17,7 +17,21 @@ name = "fluvio-run"
 path = "src/bin/main.rs"
 doc = false
 
+[features]
+default = ["telemetry"]
```
Comment (sehz) [inline src/runner/Cargo.toml:21]: let's not put this under default feature

File: src/runner/src/bin/main.rs [hunk 17]
```diff
@@ -1,10 +1,47 @@
 use structopt::StructOpt;
-use fluvio_future::task::run_block_on;
 use fluvio_run::RunCmd;
 
+use tracing::Level;
+use tracing_subscriber::{Registry, prelude::*};
+use tracing_subscriber::filter::Directive;
+
 fn main() -> Result<(), Box<dyn std::error::Error>> {
-    fluvio_future::subscriber::init_tracer(None);
```
Comment (sehz) [inline src/runner/src/bin/main.rs:6]: make jaeger optional.  if `telemetry` feature is not defined then default to previous behavior.

File: src/runner/src/bin/main.rs [hunk 18]
```diff
@@ -1,10 +1,47 @@
...
-    let cmd = RunCmd::from_args();
-    run_block_on(cmd.process())?;
+    let cmd: RunCmd = RunCmd::from_args();
+
+    #[cfg(feature = "telemetry")]
```
Comment (sehz) [inline src/runner/src/bin/main.rs:11]: use if_config to simplify

File: src/sc/src/controllers/partitions/controller.rs [hunk 19]
```diff
@@ -70,6 +70,7 @@ impl PartitionController {
         // info!("spu controller is terminated");
     }
 
+    #[instrument(skip(self, listener), name = "partition_controller_sync")]
```
Comment (sehz) [inline src/sc/src/controllers/partitions/controller.rs:73]: not fan of this renaming

File: src/sc/src/controllers/partitions/reducer.rs [hunk 20]
```diff
@@ -79,6 +79,7 @@ impl PartitionReducer {
         }
     }
 
+    #[instrument(skip(self, updates))]
```
Comment (sehz) [inline src/sc/src/controllers/partitions/reducer.rs:82]: let's put as separate PR.  One for adding feature flag.  And create separate PR for each effected modules.

**Review by sehz — CHANGES_REQUESTED** [review]
This is too big of PR.  Lets' separate into individual PR.  First should be only about putting Jaegar collector.  Then let's do individual PR for each crates

**Comment by nicholastmosher** [comment]
Minimized the PR to only include Jaeger/Opentelemetry backend. Instrumentation has been broken down by crate and I'll introduce that in further PRs.

File: src/runner/src/bin/main.rs (hunk 18 above)
Comment (nicholastmosher) [reply src/runner/src/bin/main.rs:7]: I think if_cfg would actually be about the same as this setup, and it adds an external dependency. I'd like to leave it unless you have an overwhelmingly strong opinion about it

**Review by sehz — APPROVED** [review]
Can merge after release

**Comment by sehz** [comment]
will review rest of PR once this is landed otherwise hard to see changes
